vortex-file/src/read/coalescer.rs

```diff
@@ -0,0 +1,215 @@
+#![allow(dead_code)]
+
+use std::collections::VecDeque;
+use std::io;
+use std::io::ErrorKind;
+use std::pin::Pin;
+use std::sync::{Arc, RwLock};
+use std::task::{ready, Context, Poll, Waker};
+
+use futures::Stream;
+use futures_util::future::BoxFuture;
+use futures_util::{FutureExt, StreamExt};
+use vortex_array::ArrayData;
+use vortex_error::{vortex_err, vortex_panic, VortexExpect, VortexResult};
+use vortex_io::{Dispatch, IoDispatcher, VortexReadAt, VortexReadRanges};
+
+use crate::{LayoutMessageCache, LayoutReader, Message, MessageLocator, MessageRead, RowMask};
+
+const NUM_TO_COALESCE: usize = 8;
+
+pub trait RowMaskReader<V> {
+    fn read_mask(&self, mask: &RowMask) -> VortexResult<Option<MessageRead<V>>>;
+}
+
+pub struct MaskLayoutReader {
+    layout: Box<dyn LayoutReader>,
+}
+
+impl MaskLayoutReader {
+    pub fn new(layout: Box<dyn LayoutReader>) -> Self {
+        Self { layout }
+    }
+}
+
+impl RowMaskReader<ArrayData> for MaskLayoutReader {
+    /// Read given mask out of the reader
+    fn read_mask(&self, mask: &RowMask) -> VortexResult<Option<MessageRead<ArrayData>>> {
+        self.layout.read_selection(mask)
+    }
+}
+
+enum RowMaskState<V> {
+    Pending(RowMask),
+    Ready(V),
+}
+
+pub struct Coalescer<R, S, V, RM> {
```

Comment:
Majority of interesting logic in this pr is this object. It performs multiple reads on underlying layout readers and dispatches them together. It makes sure there's ALWAYS one io request in flight. Importantly if the consumer of the stream is slow this might lead to using a lot of memory